### PR TITLE
Fix bar-less text output

### PR DESCRIPTION
**Pull request by spoutn1k**
This is my attempt at fixing #447.

The code in `draw_to_term` is a bit of mess, with double definitions, complex branching, and I tried cleaning it up. I introduce a `LineType` enum to differentiate between text, bars and empty lines. The variables names have been renamed to reflect better what they represent. Some code has been shuffled around. Tests pass and I ran examples at random with no issues.

A sample program to understand what this PR wants to fix:
```rs
static LOREM_SMALL: &str = "Lorem ipsum dolor sit amet.";
static LOREM_BIG: &str= "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec nec viverra massa. Nunc nisl lectus,
auctor in lorem eu, maximus elementum est.";

fn generate_logs(manager: &indicatif::MultiProgress, i: u32) {
    if i % 20 == 0 {
        let _ = manager.println(LOREM_BIG);
    } else if i % 10 == 0 {
        let _ = manager.println(LOREM_SMALL);
    }
}

fn main() {
    let display_manager = indicatif::MultiProgress::new();
    let bar = display_manager.add(indicatif::ProgressBar::new(100));

    for i in 0..100 {
        generate_logs(&display_manager, i);
        std::thread::sleep(std::time::Duration::from_millis(10));
    }

    for i in 0..100 {
        bar.inc(1);
        generate_logs(&display_manager, i);
        std::thread::sleep(std::time::Duration::from_millis(100));
    }
}
```

This is the output of the above before the PR:
[asciicast](https://asciinema.org/a/mwXUYBO3HC0wtYV4jLjjBVvNp)

### CODE REVIEW

File: src/draw_target.rs [hunk 1]
```diff
@@ -398,8 +398,13 @@ impl std::ops::DerefMut for DrawStateWrapper<'_> {
 impl Drop for DrawStateWrapper<'_> {
     fn drop(&mut self) {
         if let Some(orphaned) = &mut self.orphan_lines {
-            orphaned.extend(self.state.lines.drain(..self.state.orphan_lines_count));
-            self.state.orphan_lines_count = 0;
+            orphaned.extend(
```
Comment (djc) [inline src/draw_target.rs:401]: This commit has a lot going on. Would you mind trying to split it into smaller commits that make a more minimal single logical change?

File: src/draw_target.rs [hunk 2]
```diff
@@ -456,15 +461,60 @@ impl RateLimiter {
 
 const MAX_BURST: u8 = 20;
 
+#[derive(Debug, Clone, PartialEq)]
```
Comment (djc) [inline src/draw_target.rs:464]: We have a top-down order here, this should move down below any types that rely on it (`DrawState` at least).

**Review by djc — COMMENTED** [review]
Thanks for trying to improve this code. Especially because this is pretty gnarly code, it would be good to split these changes into smaller commits that make minimal (logical) changes -- for example, might introduce the `LineType` enum while representing only the types of lines that are currently in lines, and then later introduce the other variants.

**Comment by spoutn1k** [comment]
Thank you for taking the time to review this. My tests omitted the `--all-features` flag and it shows. It did not help that this PR is the result of a caffeine overdose and subsequent insomnia. Comments duly noted this PR will either be force pushed or a new one will be opened with working tests and better commits. Thanks again !

**Comment by djc** [comment]
Thanks! Force pushing is fine IME.

**Comment by spoutn1k** [comment]
https://github.com/console-rs/indicatif/blob/5396704f8e9fe781d21a5475c8d3af292d0155fe/src/draw_target.rs#L554-L556

I think my main issue stems here. For wrapping lines, this calculation is not accurate. However the concept itself seems strange to me. Why fill the line with whitespace and expect the terminal to wrap instead of moving the cursor down when needed ? This means when resizing the terminal the formatting will fall apart.

Any objections to me investigating changing this behaviour ?

**Comment by spoutn1k** [comment]
The new commit passes all the tests but two, and that is because the filler is not only printed when needed (removing one variable definition). On my machine:

```
---- multi_progress_many_bars stdout ----
thread 'multi_progress_many_bars' panicked at tests/render.rs:1230:5:
assertion failed: `(left == right)`
                                                                                                                                                                                 
Diff < left / right > :
 Clear
 Str("░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 0/10")
 Str("")
 NewLine
 Str("⠁ 0")
 Str("                                                                             ")
 Flush
 Up(1)
 Clear
 Down(1)
 Clear
 Up(1)
 Str("░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 0/10")
 Str("")
 NewLine
 Str("⠁ 0")
 Str("")
 NewLine
 Str("⠁ 1")
 Str("                                                                             ")
 Flush
 Up(2)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Up(2)
 Str("░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 0/10")
 Str("")
 NewLine
 Str("⠁ 0")
 Str("")
 NewLine
 Str("⠁ 1")
 Str("")
 NewLine
 Str("⠁ 2")
 Str("                                                                             ")
 Flush
 Up(3)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Up(3)
 Str("░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 0/10")
 Str("")
 NewLine
 Str("⠁ 0")
 Str("")
 NewLine
 Str("⠁ 1")
 Str("")
 NewLine
 Str("⠁ 2")
>Str("")
 Flush
 Up(3)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Up(3)
 Str("░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 0/10")
 Str("")
 NewLine
 Str("⠁ 0")
 Str("")
 NewLine
 Str("⠁ 1")
 Str("")
 NewLine
 Str("⠁ 2")
>Str("")
 Flush
 Up(3)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Up(3)
 Str("░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 0/10")
 Str("")
 NewLine
 Str("⠁ 0")
 Str("")
 NewLine
 Str("⠁ 1")
 Str("")
 NewLine
 Str("⠁ 2")
>Str("")
 Flush
 Up(3)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Down(1)
 Clear
 Up(3)
 Str("░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 0/10")
 Str("")
 NewLine
 Str("⠁ 0")
 Str("")
 NewLine
 Str("⠁ 1")
 Str("")
 NewLine
 Str("⠁ 2")
>Str("") <---------------------------------- Those useless prints are avoided
 Flush
```

**Comment by djc** [comment]
> I think my main issue stems here. For wrapping lines, this calculation is not accurate. However the concept itself seems strange to me. Why fill the line with whitespace and expect the terminal to wrap instead of moving the cursor down when needed ? This means when resizing the terminal the formatting will fall apart.
> 
> Any objections to me investigating changing this behaviour ?

I never have objections to other people investigating things. 😄 Suggest reading git blame to figure out why this change was made, pretty sure it wasn't just for funsies.

**Comment by djc** [comment]
> The new commit passes all the tests but two, and that is because the filler is not only printed when needed (removing one variable definition).

Removing the empty writes from the test seems fine to me.

**Comment by spoutn1k** [comment]
> Suggest reading git blame to figure out why this change was made, pretty sure it wasn't just for funsies.

I believe this is working in tandem with this check:
https://github.com/console-rs/indicatif/blob/5396704f8e9fe781d21a5475c8d3af292d0155fe/src/draw_target.rs#L549-L551

This means that for everyline line other than the first one, a newline is added. The first is expected to wrap automatically due to the filler set in my previous comment.

**Comment by spoutn1k** [comment]
This effectively fixes #447. I deeply want to keep going, because there is a lot of further refactoring that can be done, and the current solution of padding the last line will break any resizing. This is for future reference, the watermark to go back to if the PR spins out of control

**Comment by spoutn1k** [comment]
In a rare moment of lucidity I will not change the line padding in this PR. It is now ready for review again @djc !

File: src/draw_target.rs [hunk 3]
```diff
@@ -475,19 +475,37 @@ impl DrawState {
...
         &mut self,
         term: &(impl TermLike + ?Sized),
-        last_line_count: &mut VisualLines,
+        bar_count: &mut VisualLines, // The number of dynamic lines printed at the previous tick
     ) -> io::Result<()> {
         if panicking() {
             return Ok(());
         }
 
+        let term_height = term.height() as usize;
```
Comment (djc) [inline src/draw_target.rs:533]: This first commit has too many logically independent changes going on. In general, I don't think I'm a fan of hoisting up all these variable bindings, further up from where they're actually used.

Comment (spoutn1k) [reply src/draw_target.rs:533]: All the changes are variable renaming or movement. The hoisting up is to have a clear definition of what each variable is and relates to the others, especially as some of those variables were defined twice. It is also temporary and here to make the coming changes clearer.

Comment (djc) [reply src/draw_target.rs:533]: Especially for variable renaming and movement, I want those in separate commits so reviewing it is easier. I definitely don't want a whole bunch of variable declarations up front, each variable should be defined shortly before its first usage site.

File: src/draw_target.rs [hunk 4]
```diff
@@ -498,8 +498,21 @@ impl DrawState {
...
+
+        // Here we calculate the terminal vertical height each of those groups require when
+        // printing, taking wrapping into account.
+        let orphan_visual_line_count = self.visual_line_count(..text_line_count, term_width);
+        let bar_height = self.visual_line_count(text_line_count.., term_width);
+        let visual_lines = self.visual_line_count(.., term_width);
+
+        // Sanity checks
+        debug_assert!(visual_lines == orphan_visual_line_count + bar_height);
+        debug_assert!(self.orphan_lines_count <= self.lines.len());
```
Comment (djc) [inline src/draw_target.rs]: We need `visual_lines` for the `shift` definition, but I'd prefer to defer the calculation of `orphan_visual_line_count` until after the `shift` definition.

File: src/draw_target.rs [hunk 5]
```diff
@@ -471,6 +471,34 @@ pub(crate) struct DrawState {
     pub(crate) alignment: MultiProgressAlignment,
 }
 
+#[derive(Clone, Debug)]
```
Comment (djc) [inline src/draw_target.rs:475]: Please move this down to just before the `tests` module.

File: src/draw_target.rs [hunk 6]
```diff
@@ -471,6 +471,34 @@ pub(crate) struct DrawState {
...
+impl LineType {
+    pub(crate) fn inner(&self) -> &str {
+        match self {
+            LineType::Text(s) | LineType::Bar(s) => s,
+            LineType::Empty => "",
+        }
+    }
+}
+
+impl AsRef<str> for LineType {
```
Comment (djc) [inline src/draw_target.rs:507]: Let's inline `inner()` into this impl and use `as_ref()` everywhere.

File: src/draw_target.rs [hunk 7]
```diff
@@ -558,44 +587,34 @@ impl DrawState {
...
-                // This take in account linewrap from terminal
-                let terminal_len = (line_width as f64 / term_width as f64).ceil() as usize;
-
-                // If the line is effectively empty (for example when it consists
-                // solely of ANSI color code sequences, count it the same as a
-                // new line. If the line is measured to be len = 0, we will
-                // subtract with overflow later.
-                usize::max(terminal_len, 1)
-            }
-            .into();
```
Comment (djc) [inline src/draw_target.rs]: Would prefer that changes to this code happen in a separate commit from moving it into a method.

File: src/draw_target.rs [hunk 8]
```diff
@@ -667,21 +686,15 @@ impl Sub for VisualLines {
...
-        real_lines += usize::max(
-            (effective_line_length as f64 / width as f64).ceil() as usize,
-            1,
-        );
-    }
-
-    real_lines.into()
+pub(crate) fn visual_line_count(lines: &[LineType], width: usize) -> VisualLines {
+    lines.iter().fold(VisualLines::default(), |acc, line| {
+        acc.saturating_add(line.wrapped_height(width))
```
Comment (djc) [inline src/draw_target.rs:670]: Would also like the changes here to be in a separate commit.

File: src/draw_target.rs [hunk 9]
```diff
@@ -485,6 +485,22 @@ impl LineType {
             LineType::Empty => "",
         }
     }
+
+    fn console_width(&self) -> usize {
```
Comment (djc) [inline src/draw_target.rs:490]: Nit: `console_width()` should appear after `wrapped_height()`, since the latter depends on the former.

File: tests/render.rs [hunk 10]
```diff
@@ -1903,3 +1903,29 @@ fn orphan_lines_message_above_progress_bar_test(pb: &ProgressBar, in_mem: &InMem
 
     pb.finish();
 }
+
+#[test]
```
Comment (djc) [inline tests/render.rs:1907]: Nit: please move the docstring before the attribute.

**Comment by djc** [comment]
Alright, thanks for all your work on this!

**Comment by spoutn1k** [comment]
Thank you, it was a pleasure, I just am travelling a lot at the moment. Cheers !
